lib/ansible/modules/packaging/language/pip.py

```diff
@@ -523,6 +526,11 @@ def is_satisfied_by(self, version_to_test):
                 for op, ver in self._requirement.specs
             )
 
+    @staticmethod
+    def canonicalize_name(name):
+        # This is taken from PEP 503.
+        return Package._CANONICALIZE_RE.sub("-", name).lower()
```

Comment:
Which version of pip did this appear in?

Reply:
Tbh I have no idea. This is where I found it https://www.python.org/dev/peps/pep-0503/. I also found the exact code in pip, but since this a few months ago, I forgot :)

Reply:
Well, the reason I'm asking is that by using something which is present on only new envs you're adding an implicit dependency and we have to maintain backwards compatibility we committed to maintain.

Reply:
I don't think we should be referencing a private attribute/variable here in any case.  We have no guarantee that it will exist in any future release either.

Reply:
Right, that too. Recently, pip moved all of "seemingly public" modules under explicit "_internal" package to make it clear that users ain't supposed to abuse those.

Reply:
I don't understand how is this not backward compatible even if older pip versions don't canonicalize package names. For example currently on pip it's not possible to create two different packages with these names:
* `Junit-XML`
* `junit-xml`
* `junit.xml`

This is a limitation of https://pypi.org/, so you will be able to have only one of those packages. I even did a fast test by counting the number of packages, and than canonicalizing their names and counting them again:
```
# Total number of packages
$ curl -s https://pypi.org/simple/ | grep -c href=
157503

# Number of packages after canonicalization
$ curl -s https://pypi.org/simple/ | grep href= | sed -r 's#[^"]+"/simple/([^/]+)/.*#\1#g' | tr '[:upper:]' '[:lower:]' | sort | wc -l
157503
```
As you can see it's the same number of packages.

That means even if we canonicalize names on older versions, nothing will change.

Reply:
I'm talking about `ImportError`.

Reply:
@webknjaz sorry, but I don't understand about what ImportError are you talking about? I didn't use any imports in this PR as far as I can see.

Reply:
Oh, I've got an impression that you're using pip's importables. In this case it's fine.